### PR TITLE
🐛 allow for naive multi-AZ/subnet cluster creation

### LINKED ISSUE #2298: Subnet filtering does not allow for multi-az usage
/kind bug

**What steps did you take and what happened:**
I encountered this during a larger bit of research into how to achieve multi-az/subnet usage with the CAPA provider. Note that I'm attempting to use pre-created VPC and subnet info instead of the provider doing this for me.

When using something like subnet filters in an AWSMachineTemplate, only a single subnet with the tags was returned. An example of the tag-key filtering used:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: AWSMachineTemplate
metadata:
  name: wut-controlplane
  namespace: default
spec:
  template:
    spec:
...
...
      subnet:
        filters:
          - name: tag-key
            values:
              - for-capi
```

**What did you expect to happen:**
I've been scratching my brain trying to figure out a good way to make multi-az happen with this provider. My expectation was that when using filters in the method above for subnets in us-east-1[a,b,c], the selected subnet would be randomized and result in the ability to do a naive implementation of multi-az.

**Anything else you would like to add:**
The subnets are filtered and returned [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/services/ec2/instances.go#L316), which shows that subnet[0] is always returned, thus rendering the filtering kind of useless for any number of subnets tagged greater than 1.

Some further info from the slack thread about this: https://kubernetes.slack.com/archives/CD6U2V71N/p1615500253028000

**Environment:**

- Cluster-api-provider-aws version: 0.6.4
- Kubernetes version: (use `kubectl version`): 1.20.4
- OS (e.g. from `/etc/os-release`): Talos v0.9.x

### PR DESCRIPTION
**Pull request by rsmitty**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR ensures that we respect the AZ selection scheme during the
filtering of subnets. This fix can occur in two scenarios:

- The user supplies a filter for subnets which match multiple subnets
  *and* `AZSelectionScheme` is set to "Random" in the vpc spec.
- There is no subnet information specified, resulting in the default
  behavior of selecting all private subnets *and* `AZSelectionScheme`
  is set to "Random" in the vpc spec.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2298 

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by rsmitty** [comment]
WIP PR that I added mostly as a starting point to get talking/thinking about how we might solve this issue that I logged this AM. Happy to chat through any other ideas/feedback!

**Review by sedefsavas — COMMENTED** [review]
`AvailabilityZoneSelection` is for picking a AZ randomly, I think by using it out of context we are overloading its meaning here because all subnets filtered may be in the same AZ.

One other option is to add a similar field to `Filter` itself.

File: pkg/cloud/services/ec2/instances.go [hunk 1]
```diff
@@ -313,6 +319,24 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
...
 			)
 		}
+
+		// If random AZ selection is expected, attempt to return a random subnet based on the usage limit set
+		// or the number of subnets matching criteria, whichever is lower.
+		if vpcSpec.AvailabilityZoneSelection != nil && *vpcSpec.AvailabilityZoneSelection == infrav1.AZSelectionSchemeRandom {
+			rollMax := int64(*vpcSpec.AvailabilityZoneUsageLimit)
+			if len(subnets) < *vpcSpec.AvailabilityZoneUsageLimit {
+				rollMax = int64(len(subnets))
+			}
```
[the remark this answers is not in the document]
Comment (rsmitty) [reply pkg/cloud/services/ec2/instances.go:329]: From what I can tell in the code, that is only true if this is a managed VPC. Call for filtering down to the usage limit happens in a "defaultSubnets" function that's only called in managed case.

That said, your earlier comment about overloading the AvailabilityZoneSelection value is totally valid. I hadn't really considered having multiple subnets in the same AZ. Let me poke at doing this with a field to Filter and we'll see how it winds up.

**Comment by rsmitty** [comment]
@sedefsavas Curious to know what you'd think about something like my amended commit. It now adds a "FilterSelectionScheme" for AWS resources, which will allow for something like the following when setting the subnet spec in a machine template:

```  
...    
      subnet:
        filters:
          - name: tag-key
            values:
              - for-capi
        filterSelectionScheme: "Random"
```

I'm not the biggest fan of this not being scoped to _only_ subnets, but because the subnet is just an AWS resource struct, it's a bit forced upon me to do it at that level. The benefit of that, however, is that it would allow us to support randomization of other resources with filters later on if needed.

**Comment by sedefsavas** [comment]
Thanks for the changes @rsmitty, this makes sense to me. Or we could add this field to AWSMachine spec directly so that we don't add a field  that is not directly related to `AWSResourceReference`.

I will let other folks to comment on this as well to see if they'd prefer to add this as a v1alpha4 change to avoid adding it to `AWSResourceReference`. 
cc @randomvariable

**Comment by sedefsavas** [comment]
We are tracking spreading AWSMachines across multiple subnets here:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1829

This will be included in v1alpha4 changes. I'd opt for tracking this also under that issue and solve it together. WDYT @rsmitty ?

**Comment by rsmitty** [comment]
@sedefsavas Sure, we can do that. I'll join the AWS office hours next time (was out of office this week) and maybe we can chat about where we want to go.

**Comment by randomvariable** [comment]
I think we're ready to start considering this in v1alpha4 right?

**Comment by sedefsavas** [comment]
Closing this PR as well following the comment in the issue:
> We've had a bunch of discussion since this was opened and the outcome is in #2924

**Pull request closed without merging**